anchore/syft
# Expand deb cataloger to include opkg

## What?

This PR expands on the [Debian cataloger](https://github.com/anchore/syft/tree/main/syft/pkg/cataloger/deb) so that it also catalogs packages installed by [`opkg`](https://en.wikipedia.org/wiki/Opkg).

## Why?

#1847 documents the reasons that @spiffcs, myself, and others may have for wanting to be able to catalog packages installed with `opkg` using `syft`. In short, by supporting `opkg`, `syft` will support those creating SBOMs for software containerized with OpenWrt, those working in the Yocto ecosystem, and broadens support for generating SBOMs for embedded systems in general.

## How?

Because `opkg` strives to conform ["to a subset of debian’s policy manual regarding control files"](https://www.yoctoproject.org/software-item/opkg/#:~:text=Opkg%20is%20a%20lightweight%20package,as%20the%20Yocto%20Project%C2%AE.), `opkg` package metadata is actually `deb` package metadata and can be parsed with the same cataloger - albeit from a different location.

The image `openwrt/rootfs` stores `opkg` package metadata under the directory `/var/lib/opkg/info/*`. The company I work for uses tools from the Yocto ecosystem, not OpenWRT, and we have an internal image that stores the metadata under `/usr/lib/opkg/info/*`. To generalize, I made the globs look for anything under a directory like `**/lib/opkg`.

Before putting up this PR, I discussed the possibility of creating a dedicated cataloger for `opkg` with @kzantow over Slack. Not only will this not work without an upstream PR to [package-URL](https://github.com/package-url/purl-spec) to define a new type for `opkg`, but it doesn't make any sense. The `deb` PURL type is for debian *and* debian-derived packages; `opkg` packages are debian-derived. Making a separate cataloger for `opkg` is not necessary since `opkg` packages can and should be cataloged as `deb` packages.

## Testing?

1. Clone the repo with `git clone https://github.com/johnDeSilencio/syft` and navigate to the new directory with `cd syft/`
2. Checkout the most recent commit from `feature/expand-deb-cataloger-to-include-opkg`
3. Run `make bootstrap`
4. Run `make snapshot`
5. Run `syft openwrt/rootfs`. You should see something like the following output:

```none
 ✔ Loaded image
 ✔ Parsed image
 ✔ Cataloged packages      [1 packages]
NAME     VERSION  TYPE
busybox  1.36.1   binary
```

7. Run `docker run -it anchore/syft:latest openwrt/rootfs`. You should see something like the following output:

```none
 ✔ Parsed image                                                                                     sha256:df1f661107dca36745dd2bd9f36e1a760f7c90ae7b88270795d52c8b6b4697fe
 ✔ Cataloged packages              [197 packages]
NAME                        VERSION                                      TYPE
base-files                  1534-r23654-4c5d910ef1                       deb
busybox                     1.36.1                                       binary
busybox                     1.36.1-1                                     deb
ca-bundle                   20230311-1                                   deb
// ...
urngd                       2023-07-25-7aefb47b-1                        deb
usign                       2020-05-23-f1f65026-1                        deb
```
